test/zdtm/static/maps11.c

```diff
@@ -0,0 +1,246 @@
+{
+	if (!signal_jump_set)
+		return;
+
+	siglongjmp(signal_jmp_buf, 1);
+}
+
+static bool try_write_to_addr(uint8_t *ptr)
+{
+	bool failed;
+
+	/* Tell signal handler to jump back here on fatal signal. */
+	signal_jump_set = true;
+	/* If a fatal signal arose, we will jump back here and failed is set. */
+	failed = sigsetjmp(signal_jmp_buf, 1) != 0;
+
+	if (!failed)
+		*ptr = 'x';
+
+	signal_jump_set = false;
+	return !failed;
+}
+
+static int setup_sigsegv_handler(void)
+{
+	uint8_t write_me;
+
+	if (signal(SIGSEGV, handle_sigsegv) == SIG_ERR) {
+		fail("setting SIGSEGV handler failed");
+		return 1;
+	}
+
+	/* ensure that try_write_to_addr() works properly */
+	if (!try_write_to_addr(&write_me)) {
+		fail("Failed to write at valid addr. Buggy try_write_to_addr()?");
+		return 1;
+	}
+
+	if (try_write_to_addr(NULL)) {
+		fail("Failed to detect an invalid write. Buggy try_write_to_addr()?");
+		return 1;
+	}
+
+	return 0;
+}
+
+static inline void *mmap_pages(void *addr_hint, unsigned int count)
+{
+	char *map;
+
+	map = mmap(addr_hint, count * PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (map == MAP_FAILED || (addr_hint && (map != addr_hint)))
+		return MAP_FAILED;
+
+	return map;
+}
+
+static int check_guards(const char *when)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(pages); i++) {
+		if (!pages[i].guarded)
+			continue;
+
+		if (try_write_to_addr(&map_base[i * PAGE_SIZE])) {
+			fail("successful write to a guarded area %d %s C/R", i, when);
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static void gen_pages_data(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(pages); i++) {
+		uint32_t crc;
+
+		if (pages[i].guarded)
+			continue;
+
+		crc = ~0;
+		datagen(&map_base[i * PAGE_SIZE], PAGE_SIZE, &crc);
+	}
+}
+
+static int check_pages_data(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(pages); i++) {
+		uint32_t crc;
+
+		if (pages[i].guarded)
+			continue;
+
+		crc = ~0;
+		if (datachk(&map_base[i * PAGE_SIZE], PAGE_SIZE, &crc)) {
+			fail("Page %d is corrupted", i);
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static int prepare_vmas(void)
+{
+	char *map;
+	int i, shift;
+
+	shift = 0;
+	for (i = 0; i < ARRAY_SIZE(vmas); i++) {
+		map = mmap_pages(&map_base[shift * PAGE_SIZE], vmas[i].pages_num);
+		if (map == MAP_FAILED) {
+			fail("mmap of [%d,%d] pages failed", shift, shift + vmas[i].pages_num);
+			return 1;
+		}
+
+		shift += vmas[i].pages_num;
+	}
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	unsigned int pages_num = ARRAY_SIZE(pages);
+
+	test_init(argc, argv);
+
+	if (setup_sigsegv_handler())
+		return 1;
+
+	/*
+	 * ToDo:
+	 * - handle VM_WIPEONFORK in CRIU and test it too
+	 * - ...
+	 */
+
+	/* let's find a large enough area in address space */
+	map_base = mmap_pages(NULL, pages_num);
+	if (map_base == MAP_FAILED) {
+		fail("mmap of %d pages failed", pages_num);
+		return 1;
+	}
+
+	if (munmap(map_base, pages_num * PAGE_SIZE)) {
+		fail("munmap failed");
```

Comment:
nit: this is a preparation step, so it is better to use pr_err/pr_perror. fail() should be used when we check states after C/R.

Reply:
Have fixed. Thanks!